### DISCUSSION TITLE
Extend Viper env variables prefix support

### DISCUSSION THREAD

**Discussion by Oxxyg33n** [spf13/viper#1170] [Ideas]
Hi, Viper community

Another suggestion from me is to extend env variables prefix support, with another package I'm using, I'm able to parse .env files configuration with using some application prefix, let's say prefix is `EX` and my .env files contain variables like `EX_ENVIRONMENT`, but my configuration structure looks like this:

```
type Configuration struct {
   Environment string `envconfig:"environment"`
}
```

Which reads later from .env file as EX_ENVIRONMENT variable, I was expecting `viper.SetEnvPrefix("EX")` to do the same thing, but seems it only works when reading from environment variables and bindenv method. Is it possible to make it the same way so that 

```
type Configuration struct {
	Environment string `mapstructure:"environment"`
}
```

Structure above, with `viper.SetEnvPrefix("EX")` set, would read `EX_ENVIRONMENT` from .env file as I'm expecting? I'd rather do not do something like

```
type Configuration struct {
	Environment string `mapstructure:"ex_environment"`
}
```

Because when you are having 50+ microservices it's getting a bit hard, especially when the variables are being shared between services using shared embedded structures?

**Comment by sagikazarmark** [comment]
Thanks @Oxxyg33n, this is a good suggestion. I think we may already have an issue open for this.

The idea is to decouple some components from the Viper core and give more control to users (ie. let you implement your own logic for parsing env files).

I've been working on those changes for some time now. The latest batch can be found in this PR: #1167
